nopassword/forms.py

```diff
@@ -1,61 +1,133 @@
-                           "Note that it is case-sensitive."),
-        'no_cookies': _("Your Web browser doesn't appear to have cookies "
-                        "enabled. Cookies are required for logging in."),
+        'invalid_username': _(
+            "Please enter a correct %(username)s. "
+            "Note that it is case-sensitive."
+        ),
         'inactive': _("This account is inactive."),
     }
 
+    def __init__(self, *args, **kwargs):
+        super(LoginCodeRequestForm, self).__init__(*args, **kwargs)
+
+        self.username_field = get_user_model()._meta.get_field(get_user_model().USERNAME_FIELD)
+        self.fields['username'] = self.username_field.formfield()
+
+    def clean_username(self):
+        username = self.cleaned_data['username']
+
+        try:
+            user = get_user_model()._default_manager.get_by_natural_key(username)
+        except get_user_model().DoesNotExist:
+            raise forms.ValidationError(
+                self.error_messages['invalid_username'],
+                code='invalid_username',
+                params={'username': self.username_field.verbose_name},
+            )
+
+        if not user.is_active:
+            raise forms.ValidationError(
+                self.error_messages['inactive'],
+                code='inactive',
+            )
+
+        self.cleaned_data['login_code'] = models.LoginCode.create_code_for_user(user)
+
+        return username
+
+    def save(self, request, login_url=None, domain_override=None, extra_context=None):
+        login_code = self.cleaned_data['login_code']
+        login_code.next = request.GET.get('next')
+        login_code.save()
+
+        if not domain_override:
+            current_site = get_current_site(request)
+            site_name = current_site.name
+            domain = current_site.domain
+        else:
+            site_name = domain = domain_override
+
+        url = '{}://{}{}?code={}&next={}'.format(
+            'https' if request.is_secure() else 'http',
+            domain,
+            resolve_url(login_url or settings.LOGIN_URL),
+            login_code.code,
+            login_code.next,
+        )
+
+        context = {
+            'domain': domain,
+            'site_name': site_name,
+            'code': login_code.code,
+            'url': url,
+        }
+
+        if extra_context:
+            context.update(extra_context)
+
+        self.send_login_code(login_code, context)
+
+    def send_login_code(self, login_code, context, **kwargs):
+        for backend in get_backends():
+            if hasattr(backend, 'send_login_code'):
+                backend.send_login_code(login_code, context, **kwargs)
+                break
+        else:
+            raise ImproperlyConfigured(
+                'Please add a nopassword authentication backend to settings, '
+                'e.g. `nopassword.backends.EmailBackend`'
+            )
+
+
+class LoginForm(forms.Form):
+    code = forms.ModelChoiceField(
```

Comment:
Backward incompatible: We validate invalid codes, previously the result was a 404 (`get_object_or_404`)